### DISCUSSION TITLE
Fooocus 2.5.0: Enhance

### DISCUSSION THREAD

**Discussion by mashb1t** [lllyasviel/Fooocus#3281] [Show and tell]
# Fooocus 2.5.0: Enhance

This release includes a feature requested multiple times by the community (e.g. in https://github.com/lllyasviel/Fooocus/issues/3113, https://github.com/lllyasviel/Fooocus/discussions/3089, https://github.com/lllyasviel/Fooocus/discussions/3039, and a few more, also see https://github.com/lllyasviel/Fooocus/discussions/3122). Implementation by @mashb1t in [v2.5.0 (fork)](https://github.com/mashb1t/Fooocus/releases/tag/v2.5.0) / https://github.com/mashb1t/Fooocus/discussions/42 and now available in main.

## What does this feature do?

Enhance allows you to automatically upscale and/or improve parts of the picture based on either a prompt or an input image.
It is comparable to ADetailer ([repository](https://github.com/Bing-su/adetailer)), but offers better and more flexible object detection and replacement with detection and replacement prompts instead of static detection models, each having ~140MB.

### Disclaimer

It is highly recommended to use performance `Speed` or `Quality` (no performance which loads a LoRA) as any inpaint engine uses a LoRA, which may not produce the best results when combined with other performance LoRAs.
Using inpaint mode `Improve Detail (face, hand, eyes, etc.)` does not set an inpaint engine, making it compatible with all performances. The documentation of inpaint modes can be found here: https://github.com/lllyasviel/Fooocus/discussions/414
All of this is also the case for normal inpainting without enhancements.

ControlNets (`ImagePrompt`, `PyraCanny`, `CPDS`, `FaceSwap`) are currently not supported for enhance steps but can be used for image generation used as basis for enhance.

### With image generation
1. generate any image with the settings you'd like to use
2. if you're satisfied, disable random seed and open the enhance tab
2.1. (optional) Enable and define order upscaling or variation (default disabled)
2.2. Enable and configure any amount of other improvement steps.
2.3. input detection prompt (what you want to detect in the image)
2.4. input positive / negative prompt (what you want to replace the detected masks with, defaults to your normal prompts if not set)
3. generate image

### Based on an existing image
Simply open Image Input and upload an image to the Enhance tab. Prompt processing will be skipped and only enhancement steps are processed. Follow steps 2+ above.
You may set `--enable-auto-describe-image` to automatically generate the prompt after image upload.

## Examples

<details><summary>UI</summary>
<p>

![screencapture-localhost-7865-2024-06-17-00_13_50](https://github.com/lllyasviel/Fooocus/assets/9307310/f2e6f9ac-87e0-46ca-b10e-ba6d67ab8724)

</p>
</details> 

| [Original (generated)](https://gallery.mashb1t.de/gallery/-8ubXvDC-HwkxOv8Oem8cabO/vd1Wm0E8WZdNbimHmVzTXTR_) | Enhance |
|--------|--------|
| ![22732ab04379cb552d97ebf38de6](https://github.com/lllyasviel/Fooocus/assets/9307310/06aced0d-a967-4b58-ac5a-ae7fd5fbf62a) | ![image - 2024-06-17T001252 624](https://github.com/lllyasviel/Fooocus/assets/9307310/33fd2d83-73c0-4192-b99e-515c39ae853f)| 

| Original (generated)| Upscale (before) | `#1` Yellow Sundress | `#2` Hands replacement|
|--------|--------|--------|--------|
| ![342070683-260e6a0b-5f42-47a3-83fd-fc3f4310d3a7](https://github.com/user-attachments/assets/f273197c-1c12-44a9-b097-ad2e77de643f)| ![image](https://github.com/user-attachments/assets/41da80f2-7fbe-482a-b4ab-e60cf1c7954a)| ![image](https://github.com/user-attachments/assets/c8d5cbe3-f7fb-4b96-924a-795ec3ef6a0e) | ![image](https://github.com/user-attachments/assets/bb20c49f-9d34-42b6-a451-996a8c41350c)| 

## Upscaling or Variation

Before First Enhancement
- always use original prompts for upscaling or variation
- improved details when enhancing due to larger canvas size
- less context of image composition due to inpainting with 1024x1024
- recommended for enhancing smaller areas (details, faces, eyes)

After Last Enhancement
- may apply changes to the image which were already fixed by enhance
- recommended for enhancing large areas (backgrounds, clothes, whole persons)
- choose which prompt to use for upscaling or variation, use either original prompts or last filled enhancement prompts with non-empty value (increases subject recognition, prevents subject from being changed back to original prompt after enhancement. Use this when substantially changing the subject described in the original prompts)

| Original (generated)| Upscale (before) | Enhance |
|--------|--------|--------|
| ![image](https://github.com/user-attachments/assets/d6a5ff85-2766-4e5d-b025-4d3243150e13) | ![image](https://github.com/user-attachments/assets/5dd16d65-47c5-4cd7-9448-57415719f7da)| ![image](https://github.com/user-attachments/assets/94a4a975-4187-49dd-9907-648ed342281d)| 

| Original (generated)| Enhance | Upscale (after) |
|--------|--------|--------|
| ![image](https://github.com/user-attachments/assets/502894b2-55a9-42a5-b03a-ba60e5bbdcaf) | ![image](https://github.com/user-attachments/assets/2cbadb44-cdb3-4899-a998-82bdd7d519d1)| ![image](https://github.com/user-attachments/assets/45a4fed7-47da-4156-a5bc-a002bb13fe77)| 

## Models
By default it uses the SAM ([website](https://segment-anything.com/), [repository](https://segment-anything.com/)) masking model, backed by GroundingDINO ([paper](https://arxiv.org/abs/2303.05499), [diffusers docs](https://huggingface.co/docs/transformers/model_doc/grounding-dino)), but offers support for all additional models currently supported by RemBG ([repository](https://github.com/danielgatis/rembg?tab=readme-ov-file#models)). GroundingDINO + SAM do not use RemBG as handler, but have natively been implemented into focus for even better results and increased level control.

Currently supported models:

- SAM + GroundingDINO (default): A model combination which allows detection prompts and segmentation of objects in detected boxes in 3 variants ([base (default)](https://huggingface.co/mashb1t/misc/resolve/main/sam_vit_b_01ec64.pth), [large](https://huggingface.co/mashb1t/misc/resolve/main/sam_vit_l_0b3195.pth), [huge](https://huggingface.co/mashb1t/misc/resolve/main/sam_vit_h_4b8939.pth))
- u2net ([download](https://github.com/danielgatis/rembg/releases/download/v0.0.0/u2net.onnx), [source](https://github.com/xuebinqin/U-2-Net)): A pre-trained model for general use cases.
- u2netp ([download](https://github.com/danielgatis/rembg/releases/download/v0.0.0/u2netp.onnx), [source](https://github.com/xuebinqin/U-2-Net)): A lightweight version of u2net model.
- u2net_human_seg ([download](https://github.com/danielgatis/rembg/releases/download/v0.0.0/u2net_human_seg.onnx), [source](https://github.com/xuebinqin/U-2-Net)): A pre-trained model for human segmentation.
- u2net_cloth_seg ([download](https://github.com/danielgatis/rembg/releases/download/v0.0.0/u2net_cloth_seg.onnx), [source](https://github.com/levindabhi/cloth-segmentation)): A pre-trained model for Cloths Parsing from human portrait. Here clothes are parsed into 3 category: Upper body, Lower body and Full body.
- silueta ([download](https://github.com/danielgatis/rembg/releases/download/v0.0.0/silueta.onnx), https://github.com/xuebinqin/U-2-Net/issues/295): Same as u2net but the size is reduced to 43Mb.
- isnet-general-use ([download](https://github.com/danielgatis/rembg/releases/download/v0.0.0/isnet-general-use.onnx), [source](https://github.com/xuebinqin/DIS)): A new pre-trained model for general use cases.
- isnet-anime ([download](https://github.com/danielgatis/rembg/releases/download/v0.0.0/isnet-anime.onnx), [source](https://github.com/SkyTNT/anime-segmentation)): A high-accuracy segmentation for anime character.

## Tech Debt / Code Improvements
While implementing the enhance feature, multiple methods have been introduced to make code reusable and allow for iterations.
The whole async_worker.py has been restructured and is now much more clear to read and easier to use.

## Debugging

Please find debugging options in Developer Debug Mode > Inpaint:
- Debug Enhance Masks: outputs generated masks, default disabled
- Debug GroundingDINO: early return in the detection process, uses GroundingDINO boxes instead of refined SAM detections inside those boxes. This option can sometimes improve results when SAM isn't able to correctly detect inside GroundingDINO boxes. FYI: GroundingDINO supports detection prompts, SAM auto-detects objects without further promps inside of the boxes.
- GroundingDINO Box Erode or Dilate: increase/decrease the detected box area by GroundingDINO. This option can sometimes improve SAM detection when GroundingDINO accidentally cuts off elements (very rare)

| GroundingDINO | GroundingDINO + SAM |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/897ca491-6af2-4a08-b32b-52655514c765) | ![image](https://github.com/user-attachments/assets/469319c4-544f-42f6-96f0-e34ea10cedd2)|

**Comment by ErmMonzon** [comment]
Appreciate the updates and your hard work on maintaining Fooocus! Enhance works really well. Can you verify that for Pony models you still have to prompt the score? E.g score_9, score_8_up, score_7_up, score_6_up. From my tests so far it looks like you still have to. Just want to verify.

**Reply by DavidDragonsage** [reply]
The Pony preset loads the Pony style. The Pony style adds "score_9, score_8_up, score_7_up, " to the positive prompt and "score_6, score_5, score_4" to the negative prompt.

Like all other styles, you do not see these additional words on the Fooocus Settings screen. However, the "Full raw prompt" entries in the log shows that they have been added.

**Reply by ErmMonzon** [reply]
aaaahhhh! Should've looked at the log! hahaha! Appreciate the confirmation

**Comment by LuisFernando79** [comment]
May God bless you in his holy glory, thank you for your updates to my favorite tool.

**Comment by Vsatyam013** [comment]
Can i use fooocus 2.5.0 enhance in google colab, ? if yes, then how.

**Reply by mashb1t** [reply]
@Vsatyam013 You can find the link to Colab in the readme and the docs linked in Fooocus when clicking the enhance checkbox.

**Comment by alexatnv** [comment]
This is great! Thank you!

I played around and it's very cool. One thing I am wondering - if someone is wearing a t-shirt, how can I say make them wear a suit and tie? I tried a couple things but it seems confined to the bounds of the original clothing. Is there a way to segment this so that it uses more area beyond the bounds of the clothing? I tried "Below the neck and body" "Body' etc.

**Reply by mashb1t** [reply]
You can use the dilate mask option to expand the segmented mask in all directions, but it's currently not possible to just automatically generate a maks by prompt without any base inage and you're better off using normal inpainting for that.

**Comment by mzisa** [comment]
Hi, many thanks for the new release and the hard work. I am currently experimenting Enhance on an upload image (in the related section), it works very well with the face but fails while trying to enhance the body, or the torso, I mean it inpaint/enhance also the face when I don't request it explicitly in the enhance prompt. Is there a way to avoid this behavior? Am i doing any wrong steps? Thanks for your feedback

**Reply by mzisa** [reply]
Additional note: It seems working with "torso" as enhance prompt, fails with body/breasts for the body upper part

**Reply by mashb1t** [reply]
@mzisa try to experiment with different keywords as well as check if the clothing detection model (upper body) works better in this case.

**Comment by Yogesh-DevHub** [comment]
Thank you so much for the work. I see 'translate prompt' option as well but didn't find after git pull. was that discarded?

**Reply by mashb1t** [reply]
@Yogesh-DevHub translate prompt feature is available in my fork and uses google translator, which many of Fooocus' users can't use due to not having access to it (Chinese firewall, blocked by network, offline usage, etc.) => not in upstream, only in my fork.

**Comment by linplug** [comment]
Thank you for your hard work! Great product! 
But after the update, you can clearly see the border of the mask. I do not know with what it is connected, but if I do for example a mask of a person's face and I want the face to be left, and the body or clothes re-rendered can clearly see the boundary of the transition between the mask and the new image. This was not the case before the update. Can you tell me what to do to make the mask was not noticeable?  
Thank you!

**Reply by mashb1t** [reply]
@linplug the generation of images using a mask hasn't changed kn 2.5.0, but new methods of creating masks have been added.
You may dilate the mask to enlarge it (masking the face => use the dilate or erode slider for mask adjustment). You can find the slider directly in each of the enhance tabs or in advanced > advanced > developer debug mode for generating images without enhance.

Also try to experiment with different inpaint modes.

**Reply by linplug** [reply]
Bro, thanks for the quick reply! You're the best! I'll keep trying and figuring it out! =) 
1. Can you please tell me if it is possible to do color correction of the image? 
2. If we take a mask of a person's face and change only the environment, is it possible to make the mask perceive the lighting of the environment and color correction of the new environment. Just often the new image is different in color from the mask. 
If not, I think this functionality would just be fanastic. 

Bro, thank you again! Hugged you tight!

**Reply by mashb1t** [reply]
@linplug You're better off using Photoshop or any image software for this purpose, not GenAI, as GenAI tools such as Fooocus do not only change colors in an area. Color corrections without changing the image content are not easily possible in Fooocus or SDXL only. With very low denoising strength you'll not get any effect, with large denoising strength you'll basically get new faces, so you'd have to find the sweet spot.

There are some forks having Photopea included, which can provide this functionality on the same website (currently don't know the fork anymore, but saw some recently with less than 5 stars).

**Reply by EvilBT** [reply]
I also encountered this bug. Occasionally, there will be an obvious square redraw area, but the mask is the face itself, not a square mask. In theory, the redraw enhancement should not have a square redraw area with obvious square boundaries.

**Comment by builder-main** [comment]
Thank you it does work well. 

I observe a certain behaviour, when Inpainting then enhancing with Upscale or Variant, it seems that the same seed is used for all the enhance Upscaling/Variant. Meaning that when generating 2 inpaints, the 2 Upscale/Variant have the same features out of the masked area (which undergone heavy changes). I'd expected the unmasked area to get a differnt variant from the upscale factor, but they get the very same result.

Is this expected?

Inpaint + Upscale Mask area
![image](https://github.com/user-attachments/assets/fce93e00-2556-4779-9480-7967a412109a) ![image](https://github.com/user-attachments/assets/1174db1a-bb34-4654-a106-18292af8ee4c)

Inpaint + Upscale unMasked area
![image](https://github.com/user-attachments/assets/680c91d6-5be5-448b-9d99-580ad0f89fac) ![image](https://github.com/user-attachments/assets/0dc44254-c932-4668-a9ab-734d60c8480e)

**Reply by builder-main** [reply]
I believe it is beacause the Mask is also applied to the enhancement. When using Vary Strong you can see that the outside area does not vary at all. Any way to allow enhance to apply on the whole image instead of the mask only ?

**Comment by nazimbeltran-ai** [comment]
This is amazing. Just finished a course that pointed me towards Fooocus. Keep up the great work!

**Comment by melteemcaglar** [comment]
I'm new here and I'm using fooocus via pinokio. How to run mashb1t's fork with pinokio?

**Comment by kolyax** [comment]
Я тут новачок і використовую fooocus підкажіть як використовувіти api

**Reply by kolyax** [reply]
я пробував по прикладу не розумію

**Reply by kolyax** [reply]
юзаю local URL

**Reply by mashb1t** [reply]
@kolyax Please check out https://github.com/lllyasviel/Fooocus/discussions/2613

**Comment by alwadeai1** [comment]
hi i want to ask 
why when i use fooocus colab when  i dowenload any models or loras they don't show in fooocus  or don't show in the file
![Screenshot 2024-12-02 000651](https://github.com/user-attachments/assets/d8aa3710-7939-43a4-bb00-b023bcd3bb07)

![Screenshot 2024-12-01 235511](https://github.com/user-attachments/assets/7660e2ac-2a74-435e-a49f-af085ef6c04f)
![Screenshot 2024-12-01 235547](https://github.com/user-attachments/assets/98042caa-2361-4b6e-8ce0-a605ac08ffd1)

**Comment by alwadeai1** [comment]
i have problem in my code but i solv it but stel the models don't show in fooocus 
![Screenshot 2024-12-02 021741](https://github.com/user-attachments/assets/0f0a06bf-b14c-4105-834b-b1f5c0335a05)
![Screenshot 2024-12-02 021808](https://github.com/user-attachments/assets/cc188ad0-7304-414b-88d2-ae7a352fb487)

**Reply by mashb1t** [reply]
@alwadeai1 ensure to use file extensions and reload Fooocus using the "reload all" button after download.

**Reply by alwadeai1** [reply]
please could you explain more 
i didn't find "reload all " button

**Reply by alwadeai1** [reply]
and i sew you send to someone link video how run foooucs in colab and run models and loras https://github.com/lllyasviel/Fooocus/discussions/2801#discussioncomment-9210247
i do what he do in the video
thanks for help

**Reply by alwadeai1** [reply]
are you mean "reload page " button

**Reply by alwadeai1** [reply]
and how can i refresh with this button 
![Screenshot 2024-12-02 133725](https://github.com/user-attachments/assets/de0116ef-3620-414e-9647-8f29a02d48e9)

**Reply by alwadeai1** [reply]
any one poease help

**Comment by PhotographerK** [comment]
![viber_image_2025-06-21_13-21-42-630](https://github.com/user-attachments/assets/bc307194-dc34-4849-bde9-93f1304afe50)

**Comment by LuisFernando79** [comment]
Please more updates for Fooocus

**Comment by aratecorichard-cell** [comment]
I downloaded it, followed all the steps, and made sure I had a good setup to create the model; however, when I tried to reproduce the image in different scenarios, it didn't allow me. What should I do?

**Comment by fg0698714-gif** [comment]
Amigos como puedo eliminar fondo a una imagen?
